rs/types/types/src/messages/inter_canister.rs

```diff
@@ -608,9 +639,60 @@ impl RequestOrResponse {
     }
 }
 
+/// XNet message: request, response or refund.
+///
+/// The underlying request / response / refund is wrapped within an `Arc`, for
+/// cheap cloning.
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+#[cfg_attr(test, derive(ExhaustiveSet))]
+pub enum StreamMessage {
+    Request(Arc<Request>),
+    Response(Arc<Response>),
+    Refund(Arc<Refund>),
+}
+
+impl ValidateEq for StreamMessage {
+    fn validate_eq(&self, rhs: &Self) -> Result<(), String> {
+        match (self, rhs) {
+            (StreamMessage::Request(l), StreamMessage::Request(r)) => l.validate_eq(r),
+            (StreamMessage::Response(l), StreamMessage::Response(r)) => l.validate_eq(r),
+            (StreamMessage::Refund(l), StreamMessage::Refund(r)) => l.validate_eq(r),
+            _ => Err("StreamMessage enum mismatch".to_string()),
+        }
+    }
+}
+
+impl StreamMessage {
+    pub fn receiver(&self) -> CanisterId {
+        match self {
+            StreamMessage::Request(req) => req.receiver,
+            StreamMessage::Response(resp) => resp.originator,
+            StreamMessage::Refund(refund) => refund.recipient,
+        }
+    }
+
+    /// Returns the amount of cycles contained in this message.
+    pub fn cycles(&self) -> Cycles {
+        match self {
+            StreamMessage::Request(req) => req.payment,
+            StreamMessage::Response(resp) => resp.refund,
+            StreamMessage::Refund(refund) => refund.amount,
+        }
+    }
+
+    /// Returns `true` iff this is a best-effort `Response`.
+    pub fn is_best_effort_response(&self) -> bool {
+        match self {
+            StreamMessage::Response(resp) => resp.is_best_effort(),
+            _ => false,
+        }
+    }
+}
+
 /// Convenience `CountBytes` implementation that returns the same value as
-/// `RequestOrResponse::Request(self).count_bytes()`, so we don't need to wrap
-/// `self` into a `RequestOrResponse` only to calculate its estimated byte size.
+/// `RequestOrResponse::Request(self).count_bytes()` and
+/// `StreamMessage::Request(self).count_bytes()` so we don't need to wrap
```

Comment:
Is it critical that this count is correct? It looks potentially brittle if we're now relying on the two contexts staying the same forever.

Reply:
I later saw that you had a check below, so resolving this.